Fix/corl 2327 user ban bug

marcushaddon opened the pull request:
## What does this PR do?
This PR should address https://vmproduct.atlassian.net/browse/CORL-2327 by making user ban management more fine grained.

## What changes to the GraphQL/Database Schema does this PR introduce?
This introduces the `UpdateUserBanInput` input and `UpdateUserBanPayload` type, as well as the `updateUserBan(...)` mutation

## How do I test this PR?
- a "Manage User Ban" modal should now be accessible regardless of a user's status (unblocking site mods from adding additional bans to a user already banned on other sites)
- unscoped moderators should be able to add and remove bans to individual sites for a user (within the same form submission), as well as ban a user across the entire organization, and unban a user across the entire organization
- scoped moderators should be able to add additional single-site bans to users even if that user is already banned on other sites
- scoped mods will see an "All sites" option for banning, which should ban the user from all sites within that moderators scope
- scoped mods will not see a "No Sites" option (as they are not able to remove bans)
- when managing user bans on individual sites, scoped mods should be able to see all sites that a user is banned on, but not be able to unban users on any sites
 
 
## How do we deploy this PR?
I don't think any special considerations should be needed.

marcushaddon committed:
create types for UpdateUserBan mutation

marcushaddon committed:
initial implementation of updateUserBan

marcushaddon committed:
begin testing updateUserBan mutation

marcushaddon committed:
WIP working on mocking user model

marcushaddon committed:
WIP fix mocks

marcushaddon committed:
test updateUserBan

marcushaddon committed:
WIP resolving conflicts from cherry pick

marcushaddon committed:
merge develop

marcushaddon committed:
WIP cuse useEffect to set intitial val of selectedIDsInput

marcushaddon committed:
WIP

marcushaddon committed:
WIP make UserStatusSitesList reflect users banned sites

marcushaddon committed:
use UpdateUserBanMutation on form submit

marcushaddon committed:
fix UpdateUserBan (fe)
actually register updateUserBan in muation resolver (be)

marcushaddon committed:
update copy
use  as source of truth for site checkboxes

marcushaddon committed:
WIP working out display/checked/submitted issue

marcushaddon committed:
update translation for save button
fix checkbox checked state

marcushaddon committed:
fix deps array for initializing banSiteIDs

marcushaddon committed:
Merge branch 'develop' into fix/CORL-2327-user-ban-bug

marcushaddon committed:
WIP fixing tests

marcushaddon committed:
handle all site ban
do not allow scoped users to issue all site ban

marcushaddon committed:
disable manage ban button for site mods if user blanket banned

marcushaddon committed:
open manage ban modal even if blanket banned

marcushaddon committed:
(re)implement unban

marcushaddon committed:
temporarily skip failing tests

marcushaddon committed:
Merge branch 'fix/CORL-2327-user-ban-bug' of github.com:coralproject/talk into fix/CORL-2327-user-ban-bug

marcushaddon committed:
WIP reconciling ModerateCardContainer with new BanModal API

marcushaddon committed:
WIP
debugging infinite update loop in UserStatusSitesList
verifying API changes to BanModal dont break ModerateCardContainer

marcushaddon committed:
uncomment mutation

marcushaddon committed:
Merge branch 'develop' into fix/CORL-2327-user-ban-bug

marcushaddon committed:
fix existing tests

marcushaddon committed:
write test for managing individual site bans

marcushaddon committed:
filter out of scope sites from update user ban mutation

marcushaddon committed:
WIP debugging permission bug

marcushaddon committed:
hide 'no sites' for scoped users
make updateUserBan accessibile to scoped users on backend

marcushaddon committed:
WIP
disable unbanning for site mods

marcushaddon committed:
fix orgmod ban management bug

marcushaddon committed:
fix optional chaining issue
fix test

marcushaddon committed:
Merge branch 'develop' into fix/CORL-2327-user-ban-bug

marcushaddon committed:
cleanup

kabeaty commented on src/core/client/admin/components/ModerateCard/ModerateCardContainer.tsx:
```diff
@@ -275,16 +275,18 @@ const ModerateCardContainer: FunctionComponent<Props> = ({
 
   const handleBanConfirm = useCallback(
     async (
+      updateType: string,
```
It doesn't look like `updateType` and `unbanSiteIDs` are used anywhere here. Can they be removed?

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -24,12 +26,15 @@ import styles from "./BanModal.css";
 
 interface Props {
   username: string | null;
+  banStatus?: UserStatusChangeContainer_user["status"]["ban"];
```
I think ban sites are already coming through with `userScopes` so it should be possible to just get the sites that way instead of passing through again?

marcushaddon replied:
Doesn't the `Scope` type of `userScopes` refer to a moderation scope for a moderator, and `banStatus` refer to the ban status of a commenter?

kabeaty replied:
Ah, so in`UserStatusChangeContainer`, what is passed in for `userScopes.sites` is `user.status.ban.sites`. I believe only the viewer has `moderationScopes`, as they are the ones viewing the modal and banning/unbanning in this case, and then the user is the commenter who does/doesn't have a ban status and ban sites.

kabeaty commented on src/core/client/admin/components/UserStatus/UserStatusSitesList.tsx:
```diff
@@ -48,44 +52,78 @@ const UserStatusSitesList: FunctionComponent<Props> = ({ viewerScopes }) => {
-      selectedIDsInput.onChange(changed);
+  const onToggleSite = useCallback(
+    (id: string, on: boolean) => {
+      if (on) {
```
What do you think about using a ternary operator here instead for readability? Also, what do you think about using `siteID` here and below as well for consistency?

kabeaty commented on src/core/client/admin/components/UserStatus/UserStatusSitesList.tsx:
```diff
@@ -103,31 +141,57 @@ const UserStatusSitesList: FunctionComponent<Props> = ({ viewerScopes }) => {
                 <Localized id="community-banModal-allSites">
-                  <RadioButton checked={!showSites} onChange={onHideSites}>
+                  <RadioButton
+                    checked={updateType.value === "ALL_SITES"}
```
What do you think about making these updateTypes into an enum somewhere?

kabeaty commented on src/core/client/admin/test/community/community.spec.tsx:
```diff
@@ -809,20 +827,147 @@ it("remove user ban", async () => {
   expect(resolvers.Mutation!.removeUserBan!.called).toBe(true);
 });
 
+it.only("manages bans on individual sites", async () => {
```
Looks like you still need to remove the `only` here :)

kabeaty commented on src/core/client/admin/test/community/community.spec.tsx:
```diff
@@ -923,7 +1068,7 @@ it("invites user", async () => {
 });
 
-it("ban user across specific sites", async () => {
+it.skip("ban user across specific sites", async () => {
```
Did you mean to keep this test skipped? Also, this seems very similar to the test for individual sites above. Could they be combined into one test instead?

kabeaty commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -7466,6 +7466,52 @@ type BanUserPayload {
 
+input UpdateUserBanInput {
+  """
+  userID is the id of the user whos ban status should be updated
```
typo: `whos` --> `whose`

kabeaty commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -8500,7 +8546,14 @@ type Mutation {
 
   """
-  removeUserBan will remove an active ban from a User if they have one.
+  updateUserBan will ban a specified user from specified sites (if
```
`updateUserBan` is adding and removing bans, right? Can you update the comment here to clarify?

kabeaty commented on src/core/server/services/users/users.spec.ts:
```diff
@@ -0,0 +1,248 @@
+          false,
+          [siteA.id]
+        )
+    ).rejects.toThrow(); // TODO: assert correct error
```
Are these still TODOs you're looking into to assert correct errors?

kabeaty commented on src/core/server/services/users/users.spec.ts:
```diff
@@ -0,0 +1,248 @@
+  const userService = require("coral-server/models/user");
+
+  it("rejects updates by non mod users", async () => {
+    const commentor = createUserFixture({
```
nit: can you make this `commenter`?

marcushaddon committed:
Merge branch 'develop' into fix/CORL-2327-user-ban-bug

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -67,27 +73,26 @@ const BanModal: FunctionComponent<Props> = ({
+
+        const inScope = (siteID: string) =>
+          !moderationScopesEnabled ||
+          !viewerScopes?.sites?.length ||
```
Would it work to just check for `!isSiteMod` in place of lines 79 & 80 here?

marcushaddon replied:
oh thats much better!

kabeaty commented on src/core/client/admin/test/community/community.spec.tsx:
```diff
@@ -672,12 +676,14 @@ it("ban user", async () => {
   });
 
-  const popup = within(userRow).getByLabelText(
+  const statusMenu = within(userRow).getByLabelText(
```
Can you update this to be `dropdown` as well, as you did in the other tests, for consistency?

kabeaty commented on src/core/server/services/users/users.ts:
```diff
@@ -1175,6 +1175,140 @@ export async function ban(
 }
 
+/**
+ * ban will ban a specific user from interacting with Coral.
```
I think this comment should say `updateUserBan`. Also, it's possible to ban and unban with `updateUserBan`, right? I think it would be good to update to clarify that as well.

kabeaty commented on src/core/server/services/users/users.ts:
```diff
@@ -1175,6 +1175,140 @@ export async function ban(
+    banner.role !== GQLUSER_ROLE.ADMIN &&
+    banner.role !== GQLUSER_ROLE.MODERATOR
+  ) {
+    throw new Error("User not authorized to peform UpdateUserBan"); // Look for specific error
```
It looks like there's still a TODO here around the specific error?

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -67,27 +73,26 @@ const BanModal: FunctionComponent<Props> = ({
+          !viewerScopes?.sites?.length ||
+          viewerScopes?.sites?.some(({ id }) => id === siteID);
+
+        const filteredBans = banSiteIDs?.filter((siteID: string) =>
```
To me, something like `inScopeFilteredBans` would be clearer than just `filteredBans` regarding how the bans are filtered when I'm reading through. What do you think?

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -205,9 +213,13 @@ const BanModal: FunctionComponent<Props> = ({
-                        Ban User
+                    <Localized id="community-banModal-updateBan">
+                      <Button
+                        data-testid="save-ban-status"
```
Is this `data-testid` used anywhere?

kabeaty commented on src/core/client/admin/components/UserStatus/UserStatusChange.tsx:
```diff
@@ -72,37 +61,21 @@ const UserStatusChange: FunctionComponent<Props> = ({
+                  data-testid="manage-ban"
                   className={styles.dropdownButton}
-                  disabled={!onBan}
+                  // disabled={banned && viewerIsScoped} MARCUS: When can banned be true?
```
This can just be removed now, right, since the `Manage Ban` is always an option?

kabeaty commented on src/core/client/admin/components/UserStatus/UserStatusChange.tsx:
```diff
@@ -72,37 +61,21 @@ const UserStatusChange: FunctionComponent<Props> = ({
+            {
+              <Localized id="community-userStatus-manageBan">
                 <DropdownButton
+                  data-testid="manage-ban"
```
Is this `data-testid` used anywhere?

kabeaty commented on src/core/server/services/users/users.ts:
```diff
@@ -1175,6 +1175,140 @@ export async function ban(
+    await rejector.add({
+      tenantID: tenant.id,
+      authorID: targetUser.id,
+      moderatorID: banner.id, // MARCUS: is this going to reject them across all sites?
```
Is this resolved?

kabeaty commented on src/core/client/admin/components/UserStatus/UserStatus.tsx:
```diff
@@ -25,9 +25,9 @@ const render = (className: string, content: React.ReactNode) => (
 );
 
 const UserStatus: FunctionComponent<Props> = (props) => {
+  // MARCUS: possibly create bug ticket
```
Is this resolved?

kabeaty reviewed (changes requested):
Looks good overall! One issue that I noticed during manual testing was that I got an error when a user was banned on specific sites and then I tried to ban them on all sites. I got an error in this case that the user was already banned.

I'm also wondering if the text and email message should change in the `No sites` case? It still asks if the viewer is sure that they want to ban the user instead of remove bans for the user. Since it's possible to remove bans with the Specific sites option too, maybe the text should also be updated in this case too to include adding and removing bans?

marcushaddon committed:
remove comments
disable manage ban button for site mods when user is blanket banned
disable all sites option when user is blanket banned

marcushaddon committed:
Merge branch 'fix/CORL-2327-user-ban-bug' of github.com:coralproject/talk into fix/CORL-2327-user-ban-bug

marcushaddon committed:
make UpdateType a union type
handle all update types in ModerationCardContainer

marcushaddon committed:
address comments

marcushaddon committed:
use correct ValidationError type
update test

marcushaddon committed:
userScopes -> userBanStatus to be more explicit

marcushaddon committed:
handle undefined ban status

marcushaddon committed:
fix error import in test

marcushaddon committed:
update BanModal text based on UpdateType

marcushaddon committed:
commit UserStatus helpers

marcushaddon committed:
Merge branch 'develop' into fix/CORL-2327-user-ban-bug

marcushaddon committed:
hide email/comment rejection options when appropriate
fix user status display edge case

marcushaddon committed:
Merge branch 'fix/CORL-2327-user-ban-bug' of github.com:coralproject/talk into fix/CORL-2327-user-ban-bug

marcushaddon committed:
update test

marcushaddon committed:
store specific site ban state in BanModal

marcushaddon committed:
implement changes from fix/ban-modal-search

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -1,16 +1,25 @@
+} from "react";
 import { Field, Form } from "react-final-form";
 
+import { UserStatusChangeContainer_user } from "coral-admin/__generated__/UserStatusChangeContainer_user.graphql";
```
Can you move this import down below all the coral imports but above the `ModalHeader` for consistency with where it's imported in other files?

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -187,10 +248,62 @@ const BanModal: FunctionComponent<Props> = ({
+                    banActive={userBanStatus?.active}
+                    banState={[banSiteIDs, setBanSiteIDs]}
+                    unbanState={[unbanSiteIDs, setUnbanSiteIDs]}
+                    visible={
```
What do you think about just doing a check here in `BanModal` for whether to show `UserStatusSitesList` or null instead of passing it through here to check? I think that will be more readable to understand when it is/isn't shown, and more consistent with the other checks we're doing in this modal for when to show things.

kabeaty commented on src/core/client/admin/components/UserStatus/UserStatusSitesList.tsx:
```diff
@@ -31,64 +22,91 @@ export interface Scopes {
 
 interface Props {
+  visible: boolean;
+  readonly banActive?: Readonly<boolean>;
```
What do you think about just passing `banActive` and `bannedSites` through here as `userBanStatus`, too? I think it would be more readable to keep this consistent.

kabeaty commented on src/core/server/services/users/users.ts:
```diff
@@ -1175,6 +1177,157 @@ export async function ban(
+    banner.role !== GQLUSER_ROLE.MODERATOR
+  ) {
+    throw new UserForbiddenError(
+      "User not authorized to peform UpdateUserBan",
```
typo `peform` --> `perform`. Also wondering if it would be friendlier to make the error something more like `User not authorized to manage user bans.`

kabeaty commented on src/core/client/admin/components/UserStatus/UpdateUserBanMutation.ts:
```diff
@@ -0,0 +1,43 @@
+      `,
+      variables: {
+        input: {
+          ...input,
```
Can you please update to explicitly define the expected `input` here, so that we know exactly what this mutation expects?

marcushaddon committed:
fix ban modal state bug

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -58,53 +77,96 @@ const BanModal: FunctionComponent<Props> = ({
-          return { [FORM_ERROR]: "At least one site must be selected" };
+          return {
+            [FORM_ERROR]: (
+              <Localized id="specificSitesSelect-validation">
```
The error message here doesn't seem to match what's in the localization file. Do we want to update the error message to this there too? 

Also, I'm seeing this error when `All sites` and `No sites` are selected if the error was already triggered on `Specific sites` and then you navigate to them. For this reason, I think it makes sense to include the error only when `Specific sites` is selected instead of as a form error here (when you merge in my changes from PR 3804, you should be able to use something similar to how I set it up there, and it should hopefully just resolve this issue).

marcushaddon committed:
remove general ban when site banning

marcushaddon committed:
show error message if ban sites reset to original state

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -58,53 +77,99 @@ const BanModal: FunctionComponent<Props> = ({
+        const [banSiteIDs] = banIDsState;
+        const [unbanSiteIDs] = unbanIDsState;
+
+        const inScopeFilteredBans = banSiteIDs?.filter((siteID: string) =>
```
Are these checks just to be sure that ban/unban siteIDs aren't included that are out of a moderator's scope? Since they shouldn't be able to select anything out of their scope. I think that a comment could be helpful here.

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -58,53 +77,99 @@ const BanModal: FunctionComponent<Props> = ({
+    userBanStatus?.active ? UpdateType.NO_SITES : UpdateType.ALL_SITES
+  );
+
+  const banIDsState = useState<string[]>([]);
```
I'm wondering if it would work better to use `useField` in `UserStatusSitesList` in this case instead of `useState` here. That's kind of how it's being used, right? Then we could just check the input for `banIDs`/`unbanIDs` on form submit, and it wouldn't be needed to pass this state down through to `UserStatusSitesList`.

kabeaty commented on src/core/client/admin/components/ModerateCard/ModerateCardContainer.tsx:
```diff
@@ -375,26 +395,24 @@ const ModerateCardContainer: FunctionComponent<Props> = ({
+            sites: viewer.moderationScopes?.sites?.map((s) => s),
+          }}
+          moderationScopesEnabled={settings.multisite}
+          userBanStatus={comment.author.status.ban}
```
Nice renaming to better clarify what's getting passed through here!

kabeaty commented on src/core/server/models/user/user.ts:
```diff
@@ -1848,6 +1848,9 @@ export async function siteBanUser(
       $addToSet: {
         "status.ban.siteIDs": { $each: siteIDs },
       },
+      $set: {
```
Why are we setting this to false here if they are being banned?

kabeaty commented on src/locales/en-US/admin.ftl:
```diff
@@ -1235,12 +1233,18 @@ community-userStatus-changeButton =
-community-banModal-areYouSure = Are you sure you want to ban <username></username>?
-community-banModal-consequence =
+community-banModal-allSites-title = Are you sure you want to ban <username></username>?
+community-banModal-specificSites-title = Are you sure you want to manage ban status of <username></username>?
```
Maybe adding a `the` to make this `...to manage the ban status of...` would flow better?

kabeaty reviewed (changes requested):
Thanks for making the updates to this! Looking good. I just have a few comments/questions on some of the changes.

Also, it looks like the ban modal is no longer preserving the selection of rejecting existing comments / custom ban message when switching between radio buttons. Can you take a look at this?

marcushaddon committed:
fix banned sites checklist for site mod

marcushaddon committed:
change conditional display of UserStatusSitesList

marcushaddon committed:
combine bannedSites + banActive = userBanStatus

marcushaddon committed:
fix typo
make UpdateUserBanMutation input more explicit

marcushaddon committed:
improve copy
remove unnessecary filtered siteIDs

marcushaddon committed:
merge develop

marcushaddon committed:
preserve rejectComments and email preferences

marcushaddon committed:
fix sticky error display
default to SPECIFIC_SITES if user has any bans

marcushaddon committed:
WIP fixing tests

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -58,169 +72,236 @@ const BanModal: FunctionComponent<Props> = ({
+        >
+          {({
+            handleSubmit,
+            submitError,
```
Is this being used anywhere, or can it be removed?

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -58,169 +72,236 @@ const BanModal: FunctionComponent<Props> = ({
-    (input) => {
+    ({
+      updateType,
+      sendCustomMessage,
```
Is this being used anywhere, or can it be removed?

kabeaty commented on src/core/client/admin/components/UserStatus/UserStatusChangeContainer.tsx:
```diff
@@ -172,17 +164,36 @@ const UserStatusChangeContainer: FunctionComponent<Props> = ({
+  const handleUpdateBan = useCallback(
+    (updateType, rejectExistingComments, banSiteIDs, unbanSiteIDs, message) => {
+      switch (updateType) {
+        case "ALL_SITES":
```
What about using the enum here as well?

kabeaty commented on src/core/client/admin/components/UserStatus/BanModal.tsx:
```diff
@@ -58,169 +72,236 @@ const BanModal: FunctionComponent<Props> = ({
+                        <Button
+                          type="submit"
+                          ref={lastFocusableRef}
+                          disabled={
```
With the `Save` button disabled this way, if a viewer bans a user across specific sites, then all sites, and then wants to ban across the same specific sites again, they won't be able to do so, as the `Save` button will be disabled in this case. I'm not sure how often this would happen, but it would create an odd state if it does...

kabeaty commented on src/core/client/admin/components/UserStatus/UserStatusSitesList.css:
```diff
@@ -2,10 +2,6 @@
-  margin-bottom: var(--spacing-3);
-}
-
 .validationMessage {
```
Are you still using this, or should it be removed?

kabeaty reviewed (commented):
Looking good, and thanks for making changes! I have just a few more small comments on the updates.

marcushaddon committed:
Revert "WIP fixing tests"

This reverts commit 09b4926f8864679b184e69acf470010c0eef86df.

marcushaddon committed:
Revert "fix sticky error display"

This reverts commit 10e54d61c6b7a75e0c2f08fccc637e5f0164a81d.

marcushaddon committed:
Revert "preserve rejectComments and email preferences"

This reverts commit 46ef654beea17d0fbdcabb6d118ac8198e6072a2.

marcushaddon committed:
reimplemnt preserving email preferences

marcushaddon committed:
reimplement persisting reject comments preference

marcushaddon commented:
@kabeaty re: https://github.com/coralproject/talk/pull/3798#discussion_r767861411
To enable submit in this case would require that every checked site in the site ban list be submitted, which they are not. only sites  actively added by the user after selecting Specific Sites are sent with the mutation to allow for showing what sites a user is banned on, without having a site moderator submit updates to site bans for sites that are out of their scope. It's not ideal but I think it would be wiser to see how actual users interact with this component before doing more non-trivial reworking based on speculation.

marcushaddon committed:
Merge branch 'develop' into fix/CORL-2327-user-ban-bug

marcushaddon committed:
Merge branch 'fix/CORL-2327-user-ban-bug' of github.com:coralproject/talk into fix/CORL-2327-user-ban-bug

marcushaddon committed:
timeboxing and going back to useState over <Field> for broken tests

marcushaddon committed:
remove site bans on general ban

kabeaty commented:
re: https://github.com/coralproject/talk/pull/3798#issuecomment-992982273 I'm replying here for future reference that this was resolved by ensuring general bans remove site bans.

kabeaty reviewed (approved):
Looks and works great! Thanks for making the changes.
